### PR TITLE
Story sorting

### PR DESCRIPTION
**Pull request by snowystinger**
Issue: https://github.com/storybooks/storybook/issues/6439

Story sorting now takes place in story_store. This is so that the Object.values returns the correct order and sorting is not needed elsewhere.

## What I did
Added a Param to the global params called storySort which takes a sort function. If none is provided then the default alphabetical sort is used.

## How to test

- Is this testable with Jest or Chromatic screenshots? don't know yet
- Does this need a new example in the kitchen sink apps? Is showcased in react kitchen sink
- Does this need an update to the documentation? yes

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by shilman** [comment]
@snowystinger Thanks for taking a go at this. I think the issue here is that `config.js` is running in the "preview" (user code) and it serializes the options over the channel to the "manager" (storybook UI code) and the function is not serializable.

I think @ndelangen is trying to address this with a unified config that makes configuring the preview and the manager easier.

**Comment by snowystinger** [comment]
@shilman I thought that might be an issue, but got tripped up because the sorting function in config.js is getting called. It's just not returning anything. Where is @ndelangen working on this? I'd like to follow along if possible.

**Comment by ndelangen** [comment]
@snowystinger I could brief you next week if you'd like.

I could use some company, possibly some pair programming would do me some good.

Tomorrow I'll be at React Amsterdam Conference, weekend I'll be resting, or at the very least trying to rest.

**Comment by ndelangen** [comment]
What do you want to do with this PR?

**Comment by snowystinger** [comment]
@ndelangen happy to help out in any way :)
If this stuff happens in another PR, that's cool

One thought I did have was that I could take out the addParams bit and just auto sort alphabetically, then at least there'd be a deterministic way to deal with this until the sorting is exposed.

File: lib/ui/src/components/sidebar/treeview/treeview.js [hunk 1]
```diff
@@ -138,12 +138,22 @@ const Tree = props => {
   }
 };
 
+const defaultAlphabeticalSort = (a, b) => (a.id > b.id ? 1 : -1);
```
Comment (Armanio) [inline lib/ui/src/components/sidebar/treeview/treeview.js]: I advise use [localeCompare](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) for alpabetical string sort.

**Review by Armanio — CHANGES_REQUESTED** [review]

**Comment by Armanio** [comment]
Good job! Thank you! 
I left the mini notice. Check it, please.
Could you update docs?

**Comment by snowystinger** [comment]
@Armanio updated. Should I also remove the logic for pulling it from the parameters object for now?

**Review by Armanio — APPROVED** [review]

**Comment by Armanio** [comment]
No, I think everything is fine. :+1:
Thank you, good job! 
Let's wait other opinions.

**Comment by ndelangen** [comment]
Related to monoconfig

File: lib/ui/src/components/sidebar/treeview/treeview.js [hunk 2]
```diff
@@ -334,7 +351,7 @@ class TreeState extends PureComponent {
         {Filter ? <Filter key="filter" onChange={this.events.onFilter} /> : null}
         {roots.length || others.length ? (
           <Fragment>
-            {roots.map(({ id, name, children }) => (
+            {roots.sort(storySort).map(({ id, name, children }) => (
```
Comment (igor-dv) [inline lib/ui/src/components/sidebar/treeview/treeview.js]: looks like that the sort will happen on every render, IMO not the best thing to do

**Comment by tmeasday** [comment]
Sorry to weigh in late here.  I don't think we should sort the stories in the treeview (in fact we already sort the roots in the treeview which is wrong too).

The reason is that the `stories` module maintains the story hash in the core state. That hash drives all interaction with the story list, including things like *going to the next story on keyboard shortcuts*. The order of stories in that hash needs to be the source of truth, otherwise weird stuff will happen and keyboard shortcuts etc will take you to random points in the treeview.

So I think we sort the stories *once* when they come over the channel and we pick them up here: https://github.com/storybooks/storybook/blob/next/lib/api/src/modules/stories.ts#L171. 

Questions over how to sort them and how to change that as a user are more nuanced of course

File: lib/ui/src/components/sidebar/treeview/treeview.js (hunk 2 above)
Comment (snowystinger) [reply lib/ui/src/components/sidebar/treeview/treeview.js]: Agreed, had proposed it over in the issue with the same concern. Was missing the info that @tmeasday just offered up though, so now I like it even less :)
I'll try to look at the file he offered up soon.

**Comment by snowystinger** [comment]
@tmeasday @shilman updated per comments

**Comment by tmeasday** [comment]
Amazing @snowystinger!! 🙌

Let's fix `sortStoriesByKind` (#5827) in a follow up PR

**Review by tmeasday — APPROVED** [review]

**Comment by tmeasday** [comment]
@snowystinger if it is easy, you might want to drop the sorting of roots here: https://github.com/storybooks/storybook/blob/next/lib/ui/src/components/sidebar/treeview/utils.js#L62-L64 as I am guessing it is fighting against your code

**Review by shilman — CHANGES_REQUESTED** [review]
Nice fix -- super happy that this is now a responsibility of the story store 💯 

Please update the options parameters doc as well: https://github.com/storybooks/storybook/blob/next/docs/src/pages/configurations/options-parameter/index.md#L19

**Review by shilman — CHANGES_REQUESTED** [review]
@snowystinger Thanks for adding docs! Sorry for one more nitpick: any chance you can add a test to `story_store.test.js` to round this out?  Also your treeview updates look good but they broke the tests in `utils.test.js` -- can you update? Thanks for your patience on this!

**Review by shilman — APPROVED** [review]
Great work @snowystinger -- this is looking awesome. Will test it out and figure out whether to release in 5.1 and 5.2!

**Pull request merged**